### DISCUSSION TITLE
Feature Request: extend c api to support fuller set of extension apis

### DISCUSSION THREAD

**Discussion by softprops** [duckdb/duckdb#11988] [Ideas]
While it is possible to do a multitude of things with Duckdb’s extension api with c++ you can currently only do a relatively small subset of that using the c api. This limits the fuller extensibly capabilities for those outsize the c++ community. 

I learned these limitations while working on a [zig library for writing duckdb extensions](https://github.com/softprops/zig-duckdb-ext) but I believe the same limitations hold for rust and other language communities that could otherwise extend duckdb through their support for c interop, zig’s is particularly good.

For example you [can create table functions using the c api](https://duckdb.org/docs/api/c/api.html#table-functions), you can not [create scalar functions as you can with c++](https://github.com/duckdb/extension-template/blob/0714042e33ac8e30710b4146addcc2d850b6decb/src/quack_extension.cpp#L42). This extensions apis don’t exist in the c api.

This feature request would be to extend the list of extension capabilities available in the c api to match those which already exist in the c++ api so that a wider range of communities can contribute to the overall duckdb ecosystem.

**Comment by Tishj** [comment]
https://github.com/duckdb/duckdb/pull/11786

**Reply by Tishj** [reply]
Could you make a list of the features that are missing from the C API?

**Reply by softprops** [reply]
I was going to start with the unary functions but it looks like someone beat me to it! That’s great news.


I’d start next walking down the list of functions in [extensionutil](https://github.com/duckdb/duckdb/blob/main/src/main/extension/extension_util.cpp) and asking what does the equivalent look like for those using the c api.

as a thought exercise I’d be curious 
 how many of the default extensions could be implemented using the same api.

lastly I’m curious about how the motherduck extension works it’s not a function but an attach target. I believe it’s also an extension but I’m not sure what flavor class of  functions it’s registered as.

**Reply by Giorgi** [reply]
@softprops Scalar functions API is available in the feature branch: https://github.com/duckdb/duckdb/pull/11786

**Reply by softprops** [reply]
Saw that above. That’s great news!

**Comment by softprops** [comment]
@Tishj @Giorgi , I saw https://github.com/duckdb/duckdb/pull/11786 was merged and and just noticed the 1.0 release but don’t see this api. Is there a target eta/release when it can be used?

**Reply by Giorgi** [reply]
It's merged into the feature branch, not main. I don't know when it will be merged or when the next release is planned.
